### PR TITLE
Master data constrains lul

### PR DESCRIPTION
**Pull request by LucasLefevre**
Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/spreadsheet/models/spreadsheet_mixin.py [hunk 1]
```diff
@@ -21,6 +23,45 @@ class SpreadsheetMixin(models.AbstractModel):
     spreadsheet_data = fields.Text(compute='_compute_spreadsheet_data', inverse='_inverse_spreadsheet_data')
     thumbnail = fields.Binary()
 
+    @api.constrains("spreadsheet_binary_data")
+    def _check_spreadsheet_data(self):
+        for spreadsheet in self.filtered("spreadsheet_binary_data"):
+            try:
+                data = json.loads(base64.b64decode(spreadsheet.spreadsheet_binary_data).decode())
+            except (json.JSONDecodeError, UnicodeDecodeError):
+                raise ValidationError(_("Uh-oh! Looks like the spreadsheet data file isn't quite right."))
```
Comment (hokolomopo) [inline addons/spreadsheet/models/spreadsheet_mixin.py:32]: 🔥 🔥 🔥

Comment (LucasLefevre) [reply addons/spreadsheet/models/spreadsheet_mixin.py:32]: I must say I take no credit for this. It's based on an original artistic direction by a guy whose initials are F.P. and executed by Mr. GPT (firstname is Chat :cat2: )

File: addons/spreadsheet/models/spreadsheet_mixin.py [hunk 2]
```diff
@@ -21,6 +23,45 @@ class SpreadsheetMixin(models.AbstractModel):
...
+                continue
+            display_name = spreadsheet.display_name
+            errors = []
+            for model, field_chains in fields_in_spreadsheet(data).items():
+                if model not in self.env:
+                    errors.append(f"- model '{model}' used in '{display_name}' does not exist")
+                    continue
+                for field_chain in field_chains:
+                    field_model = model
+                    for fname in field_chain.split(
```
Comment (rrahir) [inline addons/spreadsheet/models/spreadsheet_mixin.py:44]: nitpick but could we have this in a single line?

Comment (LucasLefevre) [reply addons/spreadsheet/models/spreadsheet_mixin.py]: Sure.
Done

File: addons/spreadsheet/models/spreadsheet_mixin.py (hunk 1 above)
Comment (rrahir) [reply addons/spreadsheet/models/spreadsheet_mixin.py:32]: "Mister ORM, I don't feel so good"

More seriously, I would say (in this specific case) that the data file is corrupted.

File: addons/spreadsheet/models/spreadsheet_mixin.py [hunk 3]
```diff
@@ -21,6 +23,43 @@ class SpreadsheetMixin(models.AbstractModel):
...
+                record = self.env.ref(xml_id, raise_if_not_found=False)
+                if not record:
+                    errors.append(f"- xml id '{xml_id}' used in spreadsheet '{display_name}' does not exist")
+                    continue
+                # check that the menu has an action. Root menus always have an action.
+                if not record.action and record.parent_id.id:
+                    errors.append(f"- menu with xml id '{xml_id}' used in spreadsheet '{display_name}' does not have an action")
+
+            if errors:
+                raise ValidationError(_("Uh-oh! Looks like the spreadsheet data file isn't quite right.") + "\n\n" + "\n".join(errors))
```
Comment (rrahir) [inline addons/spreadsheet/models/spreadsheet_mixin.py:61]: isn't quite right. -> contains invalid data

**Comment by rrahir** [comment]
@robodoo r+

**Comment by LucasLefevre** [comment]
robodoo rebase-ff

File: addons/spreadsheet/models/spreadsheet_mixin.py [hunk 4]
```diff
@@ -21,6 +23,43 @@ class SpreadsheetMixin(models.AbstractModel):
...
+            for model, field_chains in fields_in_spreadsheet(data).items():
+                if model not in self.env:
+                    errors.append(f"- model '{model}' used in '{display_name}' does not exist")
+                    continue
+                for field_chain in field_chains:
+                    field_model = model
+                    for fname in field_chain.split("."):  # field chain 'product_id.channel_ids'
+                        if fname not in self.env[field_model]._fields:
+                            errors.append(f"- field '{fname}' used in spreadsheet '{display_name}' does not exist on model '{field_model}'")
+                        field = self.env[field_model]._fields[fname]
```
Comment (rrahir) [inline addons/spreadsheet/models/spreadsheet_mixin.py:47]: @LucasLefevre  the code crashes here, we should stop the current iteration

**Pull request closed without merging**